### PR TITLE
Add container attach

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR added container attach support in cri-containerd, and also enabled the test.

The attach code is really error prone, so please pay extra attention to see whether there is goroutine leakage or deadlock.

@kubernetes-incubator/maintainers-cri-containerd

### CODE REVIEW

File: pkg/server/io/io.go [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+	tty    bool
+	stdin  bool
+	stdout *ioutil.WriterGroup
+	stderr *ioutil.WriterGroup
+
+	closer *wgCloser
+}
+
+// WithStdin enables stdin of the container io.
+func WithStdin(stdin bool) func(*ContainerIO) error {
```
Comment (miaoyq) [inline pkg/server/io/io.go]: Define `func(*ContainerIO) error` as a type

File: pkg/server/io/io.go [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+func (c *ContainerIO) Attach(stdin io.Reader, stdout, stderr io.WriteCloser) error {
+	var wg sync.WaitGroup
+	key := util.GenerateID()
+	stdinKey := streamKey(c.id, "attach-"+key, Stdin)
+	stdoutKey := streamKey(c.id, "attach-"+key, Stdout)
+	stderrKey := streamKey(c.id, "attach-"+key, Stderr)
+
+	var stdinCloser io.Closer
+	if c.stdinPath != "" && stdin != nil {
+		f, err := fifo.OpenFifo(c.closer.ctx, c.stdinPath, syscall.O_WRONLY|syscall.O_NONBLOCK, 0700)
```
Comment (miaoyq) [inline pkg/server/io/io.go]: Make sure `c.closer != nil`

File: pkg/store/container/container.go [hunk 3]
```diff
@@ -31,18 +32,30 @@ type Container struct {
 	Status StatusStorage
 	// TODO(random-liu): Add containerd container client.
 	// TODO(random-liu): Add stop channel to get rid of stop poll waiting.
+	IO *cio.ContainerIO
+}
+
+// WithContainerIO adds IO into the container.
+func WithContainerIO(io *cio.ContainerIO) func(*Container) {
```
Comment (Random-Liu) [inline pkg/store/container/container.go]: Will Do.

File: pkg/server/io/io.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/server/io/io.go]: Will do.

File: pkg/server/io/io.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/server/io/io.go]: Good catch! Will do.

File: pkg/store/container/container.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/store/container/container.go]: Done.

File: pkg/server/io/io.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/server/io/io.go]: Done.

File: pkg/server/io/io.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/server/io/io.go]: Done.

File: pkg/server/container_start.go [hunk 4]
```diff
@@ -107,46 +107,37 @@ func (c *criContainerdService) startContainer(ctx context.Context, id string, me
...
-			if err = c.agentFactory.NewContainerLogger(logPath, agents.Stderr, stderrPipe).Start(); err != nil {
+			if stderrWC, err = cio.NewCRILogger(logPath, cio.Stderr); err != nil {
 				return fmt.Errorf("failed to start container stderr logger: %v", err)
 			}
 		}
+	} else {
+		stdoutWC = cio.NewDiscardLogger()
+		stderrWC = cio.NewDiscardLogger()
+	}
+
```
Comment (Random-Liu) [inline pkg/server/container_start.go]: Close if there is error.

**Comment by miaoyq** [comment]
@Random-Liu About `io`, I think that we can reuse the implementation in `containerd` instead of implementing it again here, Just a suggestion : )

File: pkg/server/container_attach.go [hunk 5]
```diff
@@ -17,14 +17,72 @@ limitations under the License.
...
+		return nil, fmt.Errorf("failed to find container in store: %v", err)
+	}
+	state := cntr.Status.Get().State()
+	if state != runtime.ContainerState_CONTAINER_RUNNING {
+		return nil, fmt.Errorf("container is in %s state", criContainerStateToString(state))
+	}
+	return c.streamServer.GetAttach(r)
+}
+
+func (c *criContainerdService) attach(ctx context.Context, id string, stdin io.Reader, stdout, stderr io.WriteCloser,
```
Comment (miaoyq) [inline pkg/server/container_attach.go]: Rename `attach` to `attachContainer` , like `execInContainer`,  to avoid confusion with `io.attach`

Comment (Random-Liu) [inline pkg/server/container_attach.go]: Will do.

**Comment by Random-Liu** [comment]
> @Random-Liu About io, I think that we can reuse the implementation in containerd instead of implementing it again here, Just a suggestion : )

The problem is that containerd io doesn't meet our requirement, e.g. `tee` support. I've made containerd `IO` and interface, I'll make our io implement that interface after https://github.com/kubernetes-incubator/cri-containerd/pull/113 is merged.

**Comment by miaoyq** [comment]
@Random-Liu  Got it, Thank you for your patience to explain.

**Comment by Random-Liu** [comment]
Will rebase this PR soon.

**Comment by Random-Liu** [comment]
@abhinandanpb @mikebrow Rebased, please take a look.

**Comment by Random-Liu** [comment]
@mikebrow I skipped the golang tip test, because it keeps failing even though golang 1.8.x works. And it didn't fail before, seems like a bug in golang HEAD.

Since https://github.com/kubernetes-incubator/cri-containerd/pull/145 will remove golang tip test, I'll remove it in this PR for now.

File: pkg/ioutil/write_closer_test.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+func TestWriteCloseInformer(t *testing.T) {
+	original := &writeCloser{}
+	wci, close := NewWriteCloseInformer(original)
+	data := "test"
+
+	n, err := wci.Write([]byte(data))
+	assert.Equal(t, len(data), n)
+	assert.Equal(t, data, original.buf.String())
+	assert.NoError(t, err)
+
```
Comment (mikebrow) [inline pkg/ioutil/write_closer_test.go]: maybe use another thread here to test the wait function?

File: pkg/ioutil/write_closer_test.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	wci.Close()
+	assert.True(t, original.closed)
+
+	select {
+	case <-close:
+	default:
+		assert.Fail(t, "write closer not closed")
+	}
+}
```
Comment (mikebrow) [inline pkg/ioutil/write_closer_test.go]: no-op test for the no-op write closer :-)

File: pkg/server/io/io.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+		if err != nil {
+			return err
+		}
+		// Also increase wait group here, so that `closer.Wait` will
+		// also wait for this fifo to be closed.
+		c.closer.wg.Add(1)
```
Comment (abhinandanpb) [inline pkg/server/io/io.go]: is there a corresponding Wait for this waitgroup ? Or is the closer implicitly waiting on this wg ?

File: pkg/server/container_attach.go [hunk 9]
```diff
@@ -17,14 +17,68 @@ limitations under the License.
...
+	})
+
+	// TODO(random-liu): Figure out whether we need to support historical output.
+	if err := cntr.IO.Attach(stdin, stdout, stderr); err != nil {
+		return fmt.Errorf("failed to attach container: %v", err)
+	}
+
+	// Close stdin after first attach if StdinOnce is specified.
+	if cntr.Config.StdinOnce {
+		task.CloseIO(ctx, containerd.WithStdinCloser)
```
Comment (abhinandanpb) [inline pkg/server/container_attach.go]: If StdinOnce is not specified when is the IO closed ? May be worth adding a comment as well.

File: pkg/ioutil/write_closer_test.go (hunk 6 above)
Comment (Random-Liu) [inline pkg/ioutil/write_closer_test.go]: Why another thread is required? One thread seems more deterministic here.

File: pkg/ioutil/write_closer_test.go (hunk 7 above)
Comment (Random-Liu) [inline pkg/ioutil/write_closer_test.go]: We mainly want to test `WriteCloseInformer` will close the inform channel when it is closed. And the test proves that it does. :)

File: pkg/server/io/io.go (hunk 8 above)
Comment (Random-Liu) [inline pkg/server/io/io.go]: Yeah, `ContainerIO.Close` will wait for this. This makes sure that `ContainerIO.Close` won't remove the fifo directory when Attach is still using the stdin pipe.

File: pkg/server/container_attach.go (hunk 9 above)
Comment (Random-Liu) [inline pkg/server/container_attach.go]: The IO will never close until the container dies.

Comment (Random-Liu) [inline pkg/server/container_attach.go]: Done.

**Comment by Random-Liu** [comment]
@mikebrow @abhinandanpb Replied or addressed comments.

**Comment by Random-Liu** [comment]
Apply LGTM based on https://github.com/kubernetes-incubator/cri-containerd/pull/128#pullrequestreview-58240027.

**Pull request merged**